### PR TITLE
feat: add callbacks

### PR DESCRIPTION
**Pull request by hlee9212**
## What are Callbacks?

Callbacks are **non-essential but helpful functions** that are called during training. They are like hooks, e.g. they can be called at various "timepoints" like `on_train_epoch_end` or `on_test_end`, etc. to carry out functions like:

* Saving a checkpoint (e.g. this may want to be called after each `on_ontrain_epoch_end`, for example)
* Print current metrics next to progress bar
* Log predictions
* Monitor learning rate (useful especially if the learning rate changes due to warm-up or cosine decay)
* etc.

More info: https://lightning.ai/docs/pytorch/stable/extensions/callbacks.html

## What's Added?

I have only implemented one custom Callback in `log_predictions.py` - as the name suggests, it is just to log to WandB some ground truths and predictions. 

Most are YAML files with built-in callbacks - these are in `configs/callbacks/definitions` (copied over from template repository at https://github.com/ashleve/lightning-hydra-template/tree/main). Using these pieces, I have created some presets that pick and combine these callbacks. 

Going over one example in `configs/presets/default.yaml`:
```yaml
defaults:
  - /callbacks/definitions/model_checkpoint
  - /callbacks/definitions/early_stopping
  - /callbacks/definitions/model_summary
  - /callbacks/definitions/rich_progress_bar
  - /callbacks/definitions/lr_logger
  - /callbacks/definitions/log_predictions
  - _self_

model_checkpoint:
  dirpath: ${paths.output_dir}/checkpoints
  filename: "epoch_{epoch:03d}"
  monitor: "val/loss"
  mode: "min"
  save_last: true
  auto_insert_metric_name: false

early_stopping:
  monitor: "val/loss"
  patience: 20
  mode: "min"

model_summary:
  max_depth: -1
```

This is like a compositional (😉) YAML that takes in all the elements from individual definitions under `defaults`, then modifies some arguments for some callbacks. For example, `model_checkpoint` callback is updated to use a particular filename convention, what to monitor (e.g. save when "val/loss" is minimum), etc. The `early_stopping callback` will automatically monitor "val/loss" and stop the training if it hasn't improved for 20 epochs. One could also easily change to monitor something else or save elsewhere, etc.

How callbacks are incorporated:
```python
trainer = Trainer(callbacks=[list of callbacks])
```

That's it. You will not see any Python code where I save models or check if early stopping is reached because `Trainer` from PyTorch lightning will automatically do these for you. By specifying these as callbacks, it can automatically check at "timepoints" (e.g. `on_train_epoch_end`) to save models, log predictions, whether to early stop or not, etc. The aim of PyTorch lightning is to reduce these boilerplate / helpful functions that may clutter the training logic if written fully out.

## Notes

1. This PR contains novel aspects from PyTorch Lightning that is not built into PyTorch. However, callbacks are non-essential, i.e. the code will run even if I set callbacks to `None`.
2. This PR is a good example of hierarchical configuration that `hydra` can understand.

### CODE REVIEW

File: pytorch/src/callbacks/log_predictions.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+            ]
+            gt_object_names = [
+                str(label_encoder.inverse_transform([object_ids[i].item()])[0])
+                for i in range(len(object_ids))
+            ]
+
+            # Add rows to the table
+            for i in range(len(rgbd_images)):
+                image = rgbd_images[i].cpu()  # (4, 224, 224)
+                rgb_image = image[:3, :, :]
```
Comment (nielsleadholm) [inline pytorch/src/callbacks/log_predictions.py:128]: **question:** Out of interest, did you also log the depth images at some point and check those looked reasonable?

**Review by nielsleadholm — APPROVED** [review]
Very nice, thanks for the clear summary, looks good!
